WIP - Next version of richtypo

dbismut opened the pull request:
Hey @sapegin, so here are the latest developments!

### Linting
I've enforced `eslint-config-airbnb-base` style on all files, with `prettier` compatibility. These are pretty strict rules but they make the code cleaner imo. All linting configuration is in the `.eslintrc` file. Not sure if I made the right move there so happy to make any change you would suggest.

### Rules
I think each language rules should be compiled in their own package. So I've exported a `compileRules` function from the `richtypo` package that compiles defs and rules (ie will replace all variables such as `${nbsp}` with their actual values).

### Richtypo
Ok so here is how it works now.

```js
import richtypo from 'richtypo';
import rules from 'richtypo-uk-rules'; 

const rt = richtypo(rules);
const result = rt('quotes', `This is a "texte "with inner text" in quotes"`);

console.log(
  result.string(), // removes all html tags
  result.html(), // transforms hairspace unicode into html entity &#x202F;
  result.raw(), // provides the untouched result
);

// will run all rules
rt.all('longer text');
```

So I haven't implemented the next level of currying such as `dbQuotes=rt('quotes')` but this would be easily doable).

### Example
I've rebuilt it using the latest version of `fledermaus` and made it so that it shows the original text and highlights changes to the user. Here is a capture below ;)

<img width="1052" alt="image" src="https://user-images.githubusercontent.com/5003380/47515233-9a3f7c00-d882-11e8-938f-07846bfcf1c5.png">

Notice that there is no special dash or quotes in the original text :)

That's it!

dbismut commented:
Yeah well it seems like negative lookbehind doesn't play well with Node 6. I tried to add --harmony flag to tests but that doesn't seem to work :(

sapegin commented:
Thanks! You did a lot of great work here!

> Linting
> Not sure if I made the right move there so happy to make any change you would suggest.

Almost ;-) I have my own rules (surprised they aren't used here) and a tool to add ESLint, Prettier, etc. to projects:

https://github.com/tamiadev/eslint-config-tamia/
https://github.com/sapegin/mrm

> Rules

This looks good so far.

> Richtypo

I'd prefer to return a plain string, and export conversion functions from richtypo module.

I'm now thinking that we don't really need two steps here, it could be just:

```js
import richtypo, {toPlainText} from 'richtypo';
import rules from 'richtypo-uk-rules'; 

const result = richtypo(rules.quotes, `This is a "texte "with inner text" in quotes"`);
const result = richtypo(rules.all, `This is a "texte "with inner text" in quotes"`);
// result is just a string
// to convert to text
toPlainText(result)
```

And we definitely can avoid `rt.all` and stuff ;-)

> Yeah well it seems like negative lookbehind doesn't play well with Node 6. I tried to add the --harmony flag to the test script but that doesn't seem to work :(

We can't force people to use the flag anyway ;-/

sapegin commented:
> I have my own rules (surprised they aren't used here)

It was here with Prettier and everything. Could you please revert these changes?

dbismut commented:
> It was here with Prettier and everything. Could you please revert these changes?

Sure sorry.

From your example:

```js
const result = richtypo(rules.quotes, `This is a "texte "with inner text" in quotes"`);
const result = richtypo(rules.all, `This is a "texte "with inner text" in quotes"`);
```

So we should not use rulenames anymore, correct? i.e. `richtypo(rules.quotes)` instead of `richtypo('quotes')`. We could have `rules.all` automatically return an array of all rules when compiling the ruleset.

> We can't force people to use the flag anyway ;-/

Does that mean I can remove node 6 from Travis?

sapegin commented:
> So we should not use rulenames anymore, correct? i.e. richtypo(rules.quotes) instead of richtypo('quotes'). We could have rules.all automatically return an array of all rules when compiling the ruleset.

Yep!

> Does that mean I can remove node 6 from Travis?

It means we still need to support Node 6 until April: https://github.com/nodejs/Release#release-schedule

dbismut commented:
Ok so I've made the changes.

```js
import rt from 'richtypo';
import rules from 'uk';

const { spaces, quotes, abbr, numbers, amp, emdash, ellipsis, all } = rules;

rt(spaces, `This is <b>of</b> the hook`)
// or
const allRules = rt(all);
allRules('a long text to apply all rules');
```

I wasn't able automatically create the `all` rule, so every language package will have to create it explicitly. The reason why I wasn't able to make it happen is that you might have rules that are composition of other rules, so if you would run all rules automatically, it would run composed rules twice.

Do you think it would be possible to release this version with a --beta or --next flag in npm?

dbismut commented:
I'll look into `title`, `lite`, etc. However I'm not sure if there is a rule on which composed rules should be part of title or light?

IMO there are some mainly decorative rules such as `amp` or `abbr` that should not be activated by default, but apart from them I don't see why one would want to distinguish a title from a plain text when it comes to running richtypo.

I've commented the `richtypo.js` file and also refactored some stuff again this morning so that rules can be composed in various ways:

```js
// composition with array of rules
const ruleArray = [rule1, rule2, rule3]; 

// composition with array of strings
// this requires that rule1, rule2, rule3 are also part of the ruleset
const ruleString = ['rule1', 'rule2', 'rule3'];

// composition as an object
const ruleObject = {rule1, rule2, rule3}
```

Regarding the last one, I kinda needed it only for the example and being able to run the `all` rule while knowing which composed rule it was running and decorating it through css accordingly.

Regarding the rules per se, I'm happy to discuss them as I've tweaked some. Also added a number rule and a temperature rule in addition to the French rules regarding punctuation.

I've uploaded the example site on Surge so [you can have a look](http://richtypo.surge.sh/
).

sapegin commented:
Yeah, `title` is only `amp` now, so we can get rid of it.

I'd only keep an array for simplicity:

```js
const ruleArray = [rule1, rule2, rule3]; 
```

> knowing which composed rule it was running and decorating it through css accordingly.

Can you read function `.name` instead?

> EDIT: by the way I had to eventually change the non breaking space character as it wasn't properly rendered in Safari. It's now <span style="white-space:nowrap">&thinsp;</span> 😒

Could you tell me more about that? Looks like this `span` shouldn't do anything ;-/

sapegin commented on .eslintrc:
```diff
@@ -1,3 +1,3 @@
 {
-	"extends": "tamia"
+	"extends": ["tamia", "prettier"]
```
Could you please remove Prettier here?

dbismut replied:
Done

sapegin commented:
Are you using any Yarn features like workspaces here?

dbismut commented:
> Are you using any Yarn features like workspaces here?

No, do you want me to revert on `npm`? I just find `yarn` faster but maybe that's not true anymore.

> Can you read function .name instead?

Unfortunately no. If `all` is a composition of rules as an array of plain rules, then the only name I can read is the name of the composing rule, i.e. `all`. Composing with objects (or strings) allows me to read the name from the key.

Honestly I don't think that's a big deal that we allow more composition options, is it? I think this will allow more convenient ways to build rules.

dbismut commented:
In Safari mobile it seems to render properly. No idea what happens on desktop. I've changed it back to previous `&#x202f;`.
http://richtypo.surge.sh/french

sapegin commented:
> No, do you want me to revert on npm? I just find yarn faster but maybe that's not true anymore.

In my experience npm is at least the same, but Yarn has annoying UI ;-) I'd keep npm until we need any Yarn features, which is possible for publishing multiple packages.

> Honestly I don't think that's a big deal that we allow more composition options, is it? I think this will allow more convenient ways to build rules.

I'd keep code and API as simple as possible, and any extra code needs to be tested and maintained, and creates additional cognitive load for anyone who's going to change this code.

> With some fonts and font-feature-settings or font-variant-ligatures attributes turned on in css, Safari renders a very weird

If it's fine with default settings, I'd just disable this extra settings ;-)

---

I'm just starting to look at the code and the first question is: I'm a bit puzzled why do we need compilation step, can rules be just functions that do all necessary transformations? And can defs be consts directly consumed by these functions?

```js
const nbsp = '\xA0';
const space = `[ \t${nbsp}${hairspace}]`;
const hairspace = '\xAF';

export const numberUnits = text => text.replace(
	new RegEx(`(\\d+)${space}(\\w)`),
	`$1${nbsp}$2`
);

export const temperature = text => text.replace(
	new RegExp(`(\\d)${space}?°`),
	`$1${nbsp}°`
);

export const all = [numberUnits, temperature];
```

This seems easier and more flexible. Rules can do whatever they want, aliases are just arrays of functions, an not extra compilation step.

dbismut commented:
> If it's fine with default settings, I'd just disable this extra settings ;-)

Without these settings, it renders no space at all. Which is better than the weird bug though.

---

> I'm a bit puzzled why do we need compilation step, can rules be just functions that do all necessary transformations? And can defs be consts directly consumed by these functions?

Maybe you're right and maybe I'm overcomplicating things. However, my idea was to maximize reusability of common rules. So for example the quote rule in uk replaces `"` with either `“` or `”`depending on whether it's an opening or closing quote.

In French, quotes are double quotes such as `«` or `»`. Compilation (which happens only once) allows me to import common defs and only change the `openingQuote` and `closingQuote` attributes, and import the common `quotes` rule.

Also, I'm decorating generated regexp to make sure we're not processing inside a tag.

https://github.com/sapegin/richtypo.js/blob/19363a817079c9a37105bcc33a43d01fc7d8020e/src/richtypo.js#L170

I understand we are replacing tags with numeral tags before processing the text, but you could imagining having a rule that adds css decoration to the text and add `<span class="a class">`, which could then be bought by some `quotes` or `spaces` rule.

Note that rules can still be functions, but need to be excluded from compilation (so added after). I agree this is not super elegant but still better in 99% of cases imo.

Not allowing object composition will complexify a bit the example code if I want to keep different colors on each rules. I think visualizing richtypo effects is a good advocate for the repo itself.

sapegin commented:
> However, my idea was to maximize reusability of common rules. So for example the quote rule in uk replaces " with either “ or ”depending on whether it's an opening or closing quote.

You'll have more flexibility with plain JS arrays and functions ;-) I think it's a special case and we can have a few factory functions for that:

```js
// common
const openingQutesFactory = ({ openingQuote }) => text => text.replace(
  new RegExp(`"((${tag})?(${dash}${space})?${letters})`),
  `${openingQuote}$1`
);
const closingQutesFactory = ({ closingQuote }) => text => text.replace(
  /"/,
  `${closingQuote}`
);

// en/ru/fr
export const quotes = [
  openingQutesFactory({openingQuote}),
  closingQutesFactory({closingQuote})
];
```

But I feel that even this is a premature abstraction, and a bit of duplicating for each language is perfectly fine.

> Also, I'm decorating generated regexp to make sure we're not processing inside a tag.

It also looks like a special case and not something that you need for every rule, having a const for that that can be imported and used where needed should be enough.

> Not allowing object composition will complexify a bit the example code if I want to keep different colors on each rules.

That's fine, as long as we keep the code of the actual library simple.

>  I think visualizing richtypo effects is a good advocate for the repo itself.

Sure ;-)

sapegin commented:
> I think duplicating rules would be an issue for maintenance. Rules creators (that would publish their own package) should be able to rely on common rules such as number formatting rule (which is a complex rule) without having to go into the source code and copy paste the rule. So I really think we should consider this right now.

I think you're overestimating how often this will happen ;-) Magic in the library also may change and break existing rules. In case of independent rules, it's up to the rules maintainer to update the rule or not.

> The factory idea is interesting and elegantly solves the problem above. What I'm anticipating is that if you change the signature of the common factory function for any reason (bug fix, more features), all rule packages depending on it would need to also update the rule to match the new signature?

I'm not completely sold on that idea, and this is one of the reasons. Again, there's nothing wrong with code duplication if it makes life easier ;-)

> EDIT: also if the rule is the same in a language as a common rule that has a factory function, you would still need to explicitly run the factory function against the def?

This quotes rule can't be common, because the output depends on the language.

> ...when all you could have done is:

The previous example is way easier to understand.

> By the way, it might be a good idea to create a separate package for common rules, so that it can be updated separately from richtypo?

Yup, it could include factories as well. And each language can depend on its own version of common rules packages, so it won't break if we update common rules or factories. There shouldn't be any dependencies on `richtypo` from rules I think. 

> Also there is just one last thing: in the current code I wrote, defs could be functions of each other, primarily to make them more readable. Would you create factory functions for defs also?

I couldn't understand what problem it solves, and show's it better than using another def directly ;-/

```js
const nbsp = '\xA0';
const space = `[ \t${nbsp}${hairspace}]`;
```

> To conclude, I really think RegExp are 99% of the rules that people would need for typo fixing. What brought me to your repo is how simple the json rules were looking and I'm feeling we would lose that a bit with functions.

Personally, I don't see _big_ difference in readability:

```js
export const temperature = text => text.replace(
	new RegExp(`(\\d)${space}?°`),
	`$1${nbsp}°`
);

// vs.

export const temperature = ({ space, nbsp }) => ({
	regex: `(\\d)${space}?°`,
	replace: `$1${nbsp}°`,
});
```

The function is a bit more readable, and you don't have an abstraction which is usually a good thing, — it's just JavaScript. The only reason I've done JSON, is because I wanted to share rules between JS and PHP, which never happened ;-)

> If compilation within rules package is what's bugging you we could make it happen inside richtypo, it's no big deal (and no need for memoization if we don't want to happen every time we run the rule):

The problem with compilation is that I don't see enough reasons to have it. With functions you don't need both — compilation and memoization.

dbismut commented:
Ok you know what, let me try it your way. I'll use factory functions and no compilation, we'll see how it goes.

dbismut commented:
Done! But I don't know how to highlight changes in the example file as in http://richtypo.surge.sh/ 😩

sapegin commented:
I'd highlight particular characters like non-breaking spaces. It's easier and will make the page look cleaner.

dbismut commented:
Ok done, dashes, special spaces and quotes are highlighted.

sapegin commented:
Cooool! 👍

sapegin commented:
I'd suggest to use Lerna for monorepo handling and publishing multiple packages (I have a bit of experience with that), and move all configs (Babel, .gitignore, etc.) to the root package instead of duplicating them in each package folder.

I'd also change rule packages to have a fixed prefix — `richtypo-rules-en` (not to put the language in the middle). And keep using `en` instead `uk` unless there's a strong reason to use `uk`.

sapegin commented:
And I've changed my mind about Node 6 support, let's drop it ;-)

dbismut commented:
Great! Do you want to take it from here or should I give it a try? I have no experience with Lerna but always happy to learn.

dbismut commented:
Ok well, can't say I fully understand Lerna. I moved everything into the `packages` folder including `richtypo`, moved all `devDependencies` to the root json, but I'm unsure about where to put the build and test scripts.

sapegin commented:
> I'm unsure about where to put the build and test scripts.

Tests definitely at the root, not sure about build. Would be nice to have it at the root too, if we can make it work there ;-)

dbismut commented:
Ok just pushed a new commit.

- Each package has its own `build` and `prepare` scripts that `lerna bootstrap` executes.
- Each package imports the root `babel.config.js` configuration.
- Only `example` modifies the `babel` config to add the plugin `transform-react-jsx` that is required by `fledermaus`.
- Excepts for `example`, only one `.gitignore` and `.eslintrc` files. 

I’ve also bumped node version to 7 on Babel and travis, which apparently is still failing, not sure why yet. 

There’s one thing that puzzles me thought: for `lerna bootstrap` to work, I had to move `richtypo` from `peerDependencies` to `dependencies`. Otherwise `example` would not compile. 

Other than that, I think I need to rewrite some aspects of the documentation.

sapegin commented:
That's awesome!! Thank you!

Let's move common rules to a separate package, so `richtypo` would be an independent package.

I'd also remove `flatten` because it's so easy to merge arrays with rest (`[...one, ...two]`) — less things to test and document.

Could you also use two spaces in `package.json` files as defined in the EditorConfig? That would also make diff more readable.

sapegin commented:
After this I think we can publish a beta version ;-)

dbismut commented:
> Let's move common rules to a separate package, so richtypo would be an independent package.

Ok!

> I'd also remove flatten because it's so easy to merge arrays with rest ([...one, ...two]) — less things to test and document.

Since rules can be a composition of each other, you might have deeply nested arrays right? Like below: 

```js
quotes = [openingQuote, closingQuote];
punctuation = [quotes, spaceBeforeSemiColon];
spaces = [quotes, punctuation];

all = [spaces, abbr, /* ...etc */]. 
```

So rest wouldn't suffice, would it?

sapegin commented:
You mean we're exporting rules like `quotes` and `punctuation` and want to give the use to compose them as they wish? That would make sense. In that case, I'd use Lodash instead of having custom implementation of `flatten`.

dbismut commented:
Yes, exactly: my example was a bit far-fetched, but `spaces` is already a composition of arrays exported by `common` so if you compose a rule with `spaces` you will create a nested array.

I had a look at Lodash before writing `flatten` but now that I look again, there's indeed `flattenDeep` that I had overlooked and that should do the same as `flatten`. So you're ok if we add Lodash as a the **only one** dependency right?

sapegin commented:
Or even this one: https://github.com/lukeed/arr/tree/master/packages/flatten

dbismut commented:
> Or even this one: https://github.com/lukeed/arr/tree/master/packages/flatten

Yeah well I mean it's essentially the same code, I'm just using reduce and checking if the initial argument is an array. I even simplified the function so now it looks like this:

```js
function flatten(array) {
  if (!Array.isArray(array)) {
    return [array];
  }
  return array.reduce((acc, el) => [...acc, ...flatten(el)], []);
}
```

Pushed `common` as a separate package, updated spaces in `package.json` files.

dbismut commented:
By the way, I've set node target at 8 and it compiles correctly now. In node 7 I had the following error with `fledermaus`: https://travis-ci.org/sapegin/richtypo.js/jobs/448788529#L516

sapegin commented:
I don't think we need to run examples build script on CI, but I'm not sure what exactly runs it.

dbismut commented:
Ok so maybe I've done something wrong with Lerna then.

At the moment, the `example` directory is part of `lerna.json`. To make sure all packages are built, the script `prepare` in the root `package.json` runs `lerna bootstrap` which by default runs all `prepare` scripts from all packages, and all `prepare` scripts from packages just run their local `build` script.

I could remove the `prepare` script from the `example` package so that it's not built by CI, or remove `example` from the package configuration in `lerna.json`. However, if I do so, I will lose symlinking in the `example` directory, which will force me to further configure `babel.config.js` with module resolvers (at least that's my understanding).

I would advise for the first option.

sapegin commented:
Yeah, the first option sounds better.

dbismut commented:
Looks like they do!

dbismut commented:
Ok now `example` should not be built anymore by CI.

But still negative lookbehind in regex doesn't play well with node 7 either. Actually it [doesn't work until Node 9 without the --harmony flag](https://node.green/#ES2018-features--RegExp-Lookbehind-Assertions).

For some reason CI doesn't throw an error on Node 8 (v8.12).

sapegin commented on .eslintignore:
```diff
@@ -1,2 +1,5 @@
-coverage/*
+__tests__/
+*.jsx
+packages/*/*.js
```
Looks like this will disable ESLint for most of the files ;-/

dbismut replied:
Sorry my bad. Fixed in b0c14e7231d3fa7471fbaae270b1e9db5102a46c

sapegin replied:
The pattern is still here ;-/

dbismut replied:
Ok sorry I wasn't sure which line you were referring to. Actually it only disable linting on js files at the root of packages, which are actually files built by babel. Same logic as with `.gitignore`.

sapegin commented on .prettierrc:
```diff
@@ -4,7 +4,7 @@
 	"overrides": [
 		{
-			"files": "*.md",
+			"files": ["*.md", "*.json"],
```
This will conflict with npm formatting.

dbismut replied:
Hm so maybe it’s VSCode `format on save` doing its thing but if I remove *.json from `prettierrc`overrides, then I can’t have `package.json` files rendered with spaces instead of tabs as I think you requested?

Should I replace `*.json` with `package.json`?

sapegin replied:
Nope, we shouldn't format package.json with Prettier, because npm will reformat it on install. Looks like you need to tweak your editor setup ;-/

dbismut replied:
Is it ok if I add a `.prettierignore` file then? I think a lot of people have `formatOnSave` set to `true`... See this:
https://medium.com/@martin_hotell/stop-re-formatting-package-json-with-prettier-and-vscode-once-and-for-all-52d283067f9a

sapegin replied:
I have it scoped for JS:

```json
  "[javascript]": {
    "editor.formatOnSave": true
  },
  "[javascriptreact]": {
    "editor.formatOnSave": true
  },
```

But feel free to add it to .prettierignore too.

dbismut replied:
Done in 50e4fd9f9af327d0c96afb73dcdf83114b160877

sapegin commented on .travis.yml:
```diff
@@ -1,6 +1,6 @@
 language: node_js
 node_js:
-  - 6
+  - 7
```
Node 7 isn't LTS, so it's not maintained anymore and we can remove it.

dbismut replied:
Ok! Just had a look at LTS versions and these are 6, 8, 10. 6 is out, should Travis run through 8 and 10?

sapegin replied:
Yup!

dbismut replied:
Done in 04a0e03a1a9d09bb75427b1862ba94125464f953

sapegin commented on package.json:
```diff
@@ -1,64 +1,44 @@
+    "babel-plugin-transform-react-jsx": "^6.24.1",
+    "babel-watch": "https://github.com/monathur/babel-watch.git",
+    "eslint": "^5.7.0",
+    "eslint-config-prettier": "^3.1.0",
```
I don't think we use it now.

dbismut replied:
Indeed, fixed in 95b57a52e6c8e2c6cad32cbc776529a690287185

sapegin commented on package.json:
```diff
@@ -1,64 +1,44 @@
+    "babel-core": "^7.0.0-bridge",
+    "babel-jest": "^23.6.0",
+    "babel-plugin-transform-react-jsx": "^6.24.1",
+    "babel-watch": "https://github.com/monathur/babel-watch.git",
```
Do you still use it?

dbismut replied:
Sorry, not sure what you’re referring to?

dbismut replied:
`babel-watch` should be updated soon to support babel 7 (see the issue here https://github.com/kmagiera/babel-watch/issues/75). I felt `babel-watch` was a better solution since it doesn't restart `babel-node` on every change.

sapegin replied:
OK, let's keep it then.

sapegin commented on packages/richtypo-rules-common/src/common.js:
```diff
@@ -0,0 +1,122 @@
+const space = `[ \t${nbsp}${hairspace}]`;
+const tag = '(?:<[^<>]*>)';
+const quotes = '["“”«»‘’]';
+const letters = '[a-zà-ž0-9]';
```
Should we include Cyrillic here?

dbismut replied:
Yes. Will do.

dbismut replied:
Fixed in 7e1dbd05baa43e7b937123e373c7f2b9d69b2083

sapegin commented on Readme.md:
```diff
@@ -2,116 +2,188 @@
+const text = 'There are 1000 "ways" to extend Richtypo.';
+
+// the all rule is defined in richtypo-rules-en
+// and executes all the rules from the package.
```
executes → contains? `richtypo` function executes them ;-)

dbismut replied:
I changed the Readme example text (if you have a better suggestion let me know!)
f6700a1f0b352135f7681cb48600f369a481216b

sapegin commented on packages/richtypo-rules-en/src/en.js:
```diff
@@ -0,0 +1,37 @@
+
+const numbers = [numberOrdinals, numberSeparators, numberUnits];
+
+export default {
```
Shouldn't we use named exports to be able to do:

```js
import {numbers, all} from 'richtypo-rules-en';
```

dbismut replied:
Yes you're right. Do you think we could have the `all` rule as the default export?

dbismut replied:
Done in ed79b7fd986849dcb484f5449af4cc417a8f130d

sapegin commented on .gitignore:
```diff
@@ -1,4 +1,7 @@
+coverage/
 node_modules/
+packages/*/*.js
```
And this one is weird too.

dbismut replied:
Actually packages are built with `babel` which outputs the built file at the root of the package. I don't think we should commit these files should we? I can create a `.gitignore` file ignoring `/.*js` files if you think that's clearer.

sapegin replied:
I'd add a comment ;-)

dbismut replied:
Ok done in 50e4fd9f9af327d0c96afb73dcdf83114b160877

sapegin commented:
Are you planning to do anything else here? If not I'll do a few tweaks and try to publish it.

dbismut commented:
Nope I think I’m done!

sapegin commented:
Awesome, thanks for all the work! 🤗

sapegin commented:
Interesting: I'm trying to run tests locally and all tests fail on Node 8.9.4:

>     SyntaxError: Invalid regular expression: /(?<!<[^>]*)"(((?:<[^<>]*>))?([-—][ 	 ¯])?[a-zà-ž0-9а-яё])/: Invalid group

But it passes when I upgrade to at least 8.10, which isn't mentioned on node.green.

dbismut commented:
> But it passes when I upgrade to at least 8.10, which isn't mentioned on node.green.

Indeed, that's why it passes CI Node 8 test since Travis seems to be running Node 8.12!

dbismut commented:
Also one thing: if you're running tests, I believe `fr` and `en` packages would run compiled versions of `richtypo` and `common`, so — at least in my case, I need to run `npx lerna run build:watch` in parallel to make sure `jest` re-runs tests.

sapegin commented:
Hopefully my last commit makes it unnecessary.

sapegin commented:
Nah, thanks! Just need a bit more time to finish everything.

dbismut commented:
Can't reproduce either :( This [SO](https://stackoverflow.com/questions/6288181/javascript-regex-nothing-to-repeat-error) thread seems to mention the same problem but I'm not sure I see it in the code. Did you identify the bit of code that caused the regression?

sapegin commented:
I've fixed it!

sapegin commented:
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 

Successfully published:
 - richtypo-rules-common@1.0.0-beta.0
 - richtypo-rules-en@1.0.0-beta.0
 - richtypo-rules-fr@1.0.0-beta.0
 - richtypo-rules-ru@1.0.0-beta.0
 - richtypo@4.0.0-beta.0
